thorlcr/activities/hashdistrib/thhashdistribslave.cpp

```diff
@@ -3737,36 +3737,224 @@ class HashJoinSlaveActivity : public CSlaveActivity, implements IStopInput
+                return i;
+            if (++i==htn)
+                i = 0;
+        }
+    }
+    inline void addNew(unsigned i, unsigned h, void *row, size32_t sz)
+    {
+        HTEntry *ht;
+        if (n >= ((htn * 3) / 4)) // if over 75% full
+        {
+            expand();
+            // re-find empty slot
+            i = h & (htn - 1);
+            while (true)
+            {
+                ht = &table[i];
+                if (nullptr == ht->row)
+                    break;
+                if (++i==htn)
+                    i = 0;
+            }
+        }
+        else
+            ht = &table[i];
+        ht->row = row;
+        ht->hash = h;
+        ht->size = sz;
+        n++;
+    }
+public:
+    CAggregateHT(CActivityBase &_activity, IHThorHashAggregateExtra &_extra, IHThorRowAggregator & _helper) : activity(_activity), helper(_helper)
+    {
+        hasher = _extra.queryHash();
+        elementHasher = _extra.queryHashElement();
+        comparer = _extra.queryCompareRowElement();
+        elementComparer = _extra.queryCompareElements();
+        htn = 8;
+        n = 0;
+        table = (HTEntry *)activity.queryRowManager()->allocate(((memsize_t)htn)*sizeof(HTEntry), activity.queryContainer().queryId());
+        // could check capacity and see if higher pow2
+        memset(table, 0, sizeof(HTEntry)*htn);
+    }
+    ~CAggregateHT()
+    {
+        reset();
+        ReleaseThorRow(table);
+    }
+    virtual void reset() override
+    {
+        HTEntry *t = table+htn;
+        while (t-- != table)
+        {
+            if (t->row)
+            {
+                ReleaseRoxieRow(t->row);
+                t->row = nullptr;
+            }
+        }
+        n = 0;
+        iPos = 0;
+    }
+    virtual void init(IEngineRowAllocator *_rowAllocator) override
+    {
+        rowAllocator.set(_rowAllocator);
+    }
+    virtual void addRow(const void *row) override
+    {
+        unsigned h = hasher->hash(row);
+        unsigned i = find(row, h, comparer);
+        HTEntry *ht = &table[i];
+        if (ht->row)
+        {
+            RtlDynamicRowBuilder rowBuilder(rowAllocator, ht->size, ht->row);
+            ht->size = helper.processNext(rowBuilder, row);
+            ht->row = rowBuilder.getUnfinalizedClear();
+        }
+        else
+        {
+            RtlDynamicRowBuilder rowBuilder(rowAllocator);
+            helper.clearAggregate(rowBuilder);
+            size32_t sz = helper.processFirst(rowBuilder, row);
+            addNew(i, h, rowBuilder.getUnfinalizedClear(), sz);
+        }
+    }
+    virtual void mergeElement(const void *e) override
+    {
+        unsigned h = elementHasher->hash(e);
+        unsigned i = find(e, h, elementComparer);
+        RtlDynamicRowBuilder rowBuilder(rowAllocator);
+        size32_t sz;
+        HTEntry &ht = table[i];
+        if (ht.row)
+        {
+            RtlDynamicRowBuilder rowBuilder(rowAllocator, ht.size, ht.row);
```

Comment:
ht.size isn't actually the same as maxLength, but I don't think it will cause any ill effects.